### DISCUSSION TITLE
Switching to a session from a disallowed directory disables the session

### DISCUSSION THREAD

**Discussion by oldenj** [olimorris/persisted.nvim#122] [Q&A] [answered]
Hi,
first of all, thanks for this great plugin. I have noticed the following behaviour and I'm not sure whether it's a bug:

1. Have an existing session for e.g. `~/projects/project.git/`
2. Have only `~/projects` in `allowed_dirs`, `autload = true` and `autosave = true`
3. Open neovim from a directory not present in 'allowed_dirs', e.g. `~/`
4. Do ":Telescope persisted" and select the existing session for `~/projects/project.git`
5. `vim.g.persisting` will be `nil` and any changes to the session will not be saved

If I now add `~` to `allowed_dirs` and repeat steps 1. to 4., `vim.g.persisting` is `true` and everything works as expected.

**Comment by oldenj** [comment]
I have looked through the code and my understanding is this:

1. The plugin always calls `M.save()` on `VimLeavePre`.  On this call, the session will be saved if it is "running", i.e. `vim.g.persisting == true`, autosave is enabled and the `should_autosave` callback does not evaluate to `false`.
2. When the plugin initialises via `M.setup()`, the session is set to "running", i.e. `vim.g.persisting = true`, if autosaving is enabled and the various criteria are fulfilled (directory is allowed and so on).
3. When the user loads an existing session via `:SessionLoad`, i.e. `M.load()`, the session is similarly set to "running" after the session has been loaded with `utils.load_session`.
4. When the user loads a session via Telescope, `utils.load_session()` is called directly and and the session will never be set to "running" - this isn't a problem if neovim was started from an allowed directory, because in this case the session was already set to "running" by the setup mechanism (2.).

A simple first idea could be to generalise `M.load()` such that it can be called from the Telescope action, instead of the Telescope action using `utils.load_session` directly.

@olimorris I would appreciate your input as per the contributing guidlines. I would be happy to implement a solution.

**Reply by olimorris** [reply, marked as answer]
Spot on.

I kept the two load functions separate as the Telescope action is much simpler than when a user may call `M.load()` in the plugins workflow.

More than happy to accept PRs that simplify the code. `init.lua` is a mess imo 😆
